pkg/nvlib/device/identifier.go

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package device
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/google/uuid"
+)
+
+// Identifier represents a GPU device in the form of an index or UUID
+type Identifier string
+
+// IsGpuIndex checks if an identifier is a full GPU index
+func (i Identifier) IsGpuIndex() bool {
+	if _, err := strconv.ParseUint(string(i), 10, 0); err != nil {
```

Comment:
Do we want a:
```
func (i Identifier) ToIndex() (int, error) {

}
```
that encapsulate the parseing? It seems as if we'd anyway need this in our code for using the Id as an index (i.e. passing it to `GetDeviceByIndex`).

Reply:
What about some public helpers that are not a member of `Identifier`?

```
func ParseGpuIndex(s string) (int, error) {
    i := Identifier(s)
    if !i.IsGpuIndex() {
        return 0, fmt.Errorf("not a GPU index")
    }
    return strconv.Atoi(s)
}

func ParseMigIndex(s string) (int, int, error) {
    i := Identifier(s)
    if !i.IsMigIndex() {
        return 0, 0, fmt.Errorf("not a MIG device index")
    }
    split := strings.Split(s, ":")
    gpuIdx, _ := strconv.Atoi(split[0])
    migIdx, _ := strconv.Atoi(split[1])
    return gpuIdx, migIdx, nil
}
```

And then our client side simplifies to
```
if idx, err := nvdevice.ParseGpuIndex(id); err == nil {
    return nvmllib.DeviceGetHandleByIndex(idx)
}

if gpuIdx, migIdx, err := nvdevice.ParseMigIndex(id); err == nil {
    if parent, err = l.nvmllib.DeviceGetHandleByIndex(gpuIdx); err != nvml.SUCCESS {
        return nil, fmt.Errorf("failed to get parent device handle: %w", err)
    }
    return parent.GetMigDeviceHandleByIndex(migIdx)
}

return nil, fmt.Errorf("identifier is not a valid UUID or index: %q", id)
```

Reply:
Or instead of public helpers, should this be added to the deviceLib API? https://github.com/NVIDIA/go-nvlib/blob/main/pkg/nvlib/device/api.go#L24

Reply:
I think for the purpose of this PR, we could have these as private helpers. We could then look to migrate them once we have more input on usage.

I can see something like `ParseDeviceId(string) (Identifier, error)` being useful going forward.